src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1106 @@
+            delete pkt;
+            return false;
+        }
+
+        Tick now = curTick();
+        waitingForMem.sample(now - queue_time);
+
+        _dataPort.sendPacket(pkt);
+
+        if (req->isLLSC()) {
+            TheISA::handleLockedRead(tc, req);
+        }
+        DPRINTF(FlexCPUCoreEvent, "Sent a memory request(pa: %#x)\n",
+                                req->getPaddr());
+
+        // When the result comes back from memory, call completeMemAccess
+        outstandingMemReqs.emplace(pkt, [this, pkt, inst, context, trace_data,
+                                         callback_func, now]
+        {
+            completeMemAccess(pkt, inst, context, trace_data, callback_func,
+                              now);
+        });
+
+        return true;
+    });
+
+    memoryUnit.schedule();
+
+    return true;
+}
+
+bool
+FlexCPU::requestMemWrite(const RequestPtr& req, ThreadContext* tc,
+                                   StaticInstPtr inst,
+                                   weak_ptr<ExecContext> context,
+                                   Trace::InstRecord* trace_data,
+                                   uint8_t* data, MemCallback callback_func)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestMemWrite()\n");
+
+    // Request can be handled immediately
+    if (req->getFlags().isSet(Request::NO_ACCESS)) {
+        return false; // TODO handle now.
+    }
+
+    PacketPtr pkt = Packet::createWrite(req);
+    // Note: pkt is deleted in DataPort::recvTimingResp();
+
+    if (data) {
+        const unsigned req_size = req->getSize();
+
+        pkt->allocate();
+        memcpy(pkt->getPtr<uint8_t>(), data, req_size);
+    } else { // Assume that no pointer given means we're filling with zero
+        assert(req->getFlags() & Request::STORE_NO_DATA);
+        // Maybe this line is important? Not all CPUs make this check, so I'm
+        // leaving this commented out to see if it causes problems...
+        // memset(mailbox, 0, req_size);
+        panic("Should be unreachable...");
+    }
+
+    // Alpha things?
+    if (req->isMmappedIpr()) {
+        // TODO think about limiting access to mmapped IPR
+        panic("FlexCPU not programmed to understand mmapped IPR!");
+        // Cycles delay = TheISA::handleIprWrite(tc, pkt);
+        // TODO create an event delay cycles later that calls completeAcc() on
+        //      the StaticInst.
+        return false;
+    }
+
+    Tick queue_time = curTick();
+
+    memoryUnit.addRequest([this, req, tc, inst, context, trace_data,
+                           callback_func, pkt, queue_time]
+    {
+        if (context.expired()) {
+            // No need to send req if squashed
+            delete pkt;
+            return false;
+        }
+
+        Tick now = curTick();
+        waitingForMem.sample(now - queue_time);
+
+        if (req->isLLSC()) {
+            if (!TheISA::handleLockedWrite(tc, req, cacheBlockMask)) {
+                panic("Haven't quite decided how to do this yet");
```

Comment:
Should you decide how to do this, or is it an error?

Reply:
I guess I never really finished implementing LLSC, since we never came across a workload that used it?